lib/terminus.js

```diff
@@ -1,143 +1,179 @@
-        await beforeShutdown()
-        await asyncServerStop()
-        await onSignal()
-        await onShutdown()
-        signals.forEach(sig => process.removeListener(sig, cleanup))
-        process.kill(process.pid, signal)
-      } catch (error) {
-        logger('error happened during shutdown', error)
-        process.exit(1)
-      }
+    if (statusResponse.status === 'ok') {
+      res.statusCode = 200
+    } else {
+      res.statusCode = 503
     }
+
+    return res.end(JSON.stringify(statusResponse))
   }
-  signals.forEach(
-    sig => process.on(sig, cleanup)
-  )
-}
 
-function terminus (server, options = {}) {
-  const { signal = 'SIGTERM',
-    signals = [],
-    timeout = 1000,
-    healthChecks = {},
-    onSendFailureDuringShutdown,
-    onShutdown = noopResolves,
-    beforeShutdown = noopResolves,
-    logger = noop } = options
-  const onSignal = options.onSignal || options.onSigterm || noopResolves
-  const state = Object.assign({}, intialState)
-
-  if (Object.keys(healthChecks).length > 0) {
-    decorateWithHealthCheck(server, state, {
-      healthChecks,
-      logger,
-      onSendFailureDuringShutdown
-    })
+  _decorateWithHealthCheck () {
+    this._server.listeners('request')
+      .forEach((listener) => {
+        this._server.removeListener('request', listener)
+        this._server.on('request', async (req, res) => this._onRequest(req, res, listener))
+      })
+  }
+
+  _decorateWithSignalHandler () {
+    stoppable(this._server, this._options.timeout)
+
+    this._asyncServerStop = promisify(this._server.stop).bind(this._server)
+    this._options.signals.forEach(
+      sig => {
+        process.on(sig, () => this._cleanup())
+      }
+    )
+  }
+
+  /**
+   * Returns the http server
+   */
+  get server () {
+    return this._server
   }
 
-  // push the signal into the array
-  // for backwards compatability
-  if (!signals.includes(signal)) signals.push(signal)
-  decorateWithSignalHandler(server, state, {
-    signals,
-    onSignal,
-    beforeShutdown,
-    onShutdown,
-    timeout,
-    logger
-  })
-
-  return server
+  /**
+   * Returns the status of the health check
+   * of the given url
+   * @param {string} url The url of the health check
+   * @throws {HealthCheckNotFoundError} Will throw an error if the given url
+   * has no registered health checks
+   */
+  async getHealthStatus (url) {
+    if (!this._options.healthChecks[url]) {
+      throw new HealthCheckNotFoundError('Given health check is not registered')
+    }
+
+    if (this._state.isShuttingDown) {
+      const healthRespose = this._getErrorResponse()
+      if (healthRespose !== 'ok') {
+        await this._options.onSendFailureDuringShutdown()
+      }
+      return healthRespose
+    }
+
+    let info
+
+    try {
+      info = await this._executeHealtCheck(url)
+    } catch (error) {
+      this._options.logger('healthcheck failed', error)
+      return this._getErrorResponse(error.causes || [error.message])
```

Comment:
In case ANY error is caused during health checks, this will catch it and display the error message. Is this a good idea security wise, or maybe we should only catch error messages instance of `TerminusError`?